### PR TITLE
Inertia visual

### PR DESCRIPTION
**Pull request by atharva-18**
# 🎉 New feature

## Summary
Adds inertia visual to debug models.

Screenshot from 2021-06-03 12-23-56

## Test it
I have added a sample called `visualization_demo` in the examples folder.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: include/ignition/rendering/InertiaVisual.hh [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    /// \brief Represents a inertia visual
+    class IGNITION_RENDERING_VISIBLE InertiaVisual :
+      public virtual Visual
+    {
+      /// \brief Destructor
+      public: virtual ~InertiaVisual() {}
+
+      /// \brief Load the Inertia visual from its pose and scale
+      /// \param[in] _pose Pose of the Inertia visual
+      /// \param[in] _scale scale factor of the box visual
```
Comment (ahcorde) [inline include/ignition/rendering/InertiaVisual.hh:42]:
```suggestion
      /// \param[in] _scale scale factor of the box visual
```
nit: Upper case

File: ogre/include/ignition/rendering/ogre/OgreInertiaVisual.hh [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+      // Documentation inherited.
+      public: virtual void PreRender() override;
+
+      // Documentation inherited.
+      public: Ogre::MovableObject *OgreObject() const;
+
+      /// \brief Load the Inertia visual from its pose and scale
+      /// \param[in] _pose Pose of the Inertia visual
+      /// \param[in] _scale scale factor of the box visual
```
Comment (ahcorde) [inline ogre/include/ignition/rendering/ogre/OgreInertiaVisual.hh:62]:
```suggestion
      /// \param[in] _scale Scale factor of the box visual
```
nit: Upper case

File: ogre/include/ignition/rendering/ogre/OgreInertiaVisual.hh [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+          const ignition::math::Vector3d &_scale) override;
+
+      // Documentation inherited.
+      public: virtual MaterialPtr Material() const;
+
+      // Documentation inherited.
+      public: virtual void SetMaterial(
+        MaterialPtr _material, bool _unique) override;
+
+      /// \brief Set material to grid geometry.
```
Comment (ahcorde) [inline ogre/include/ignition/rendering/ogre/OgreInertiaVisual.hh:73]: Review comment

File: ogre/src/OgreInertiaVisual.cc [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  this->SetMaterialImpl(derived);
+}
+
+//////////////////////////////////////////////////
+void OgreInertiaVisual::SetMaterialImpl(OgreMaterialPtr _material)
+{
+  std::string materialName = _material->Name();
+  Ogre::MaterialPtr ogreMaterial = _material->Material();
+
+#if (OGRE_VERSION <= ((1 << 16) | (10 << 8) | 7))
```
Comment (ahcorde) [inline ogre/src/OgreInertiaVisual.cc:144]: Can you add a comment here indicating the version of Ogre in this `#if` ?

File: ogre2/include/ignition/rendering/ogre2/Ogre2InertiaVisual.hh [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+      // Documentation inherited.
+      public: virtual void Init() override;
+
+      // Documentation inherited.
+      public: virtual void PreRender() override;
+
+      /// \brief Load the Inertia visual from its pose and scale
+      /// \param[in] _pose Pose of the Inertia visual
+      /// \param[in] _scale scale factor of the box visual
```
Comment (ahcorde) [inline ogre2/include/ignition/rendering/ogre2/Ogre2InertiaVisual.hh:57]:
```suggestion
      /// \param[in] _scale Scale factor of the box visual
```
nit: Upper case

File: ogre2/include/ignition/rendering/ogre2/Ogre2InertiaVisual.hh [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+          const ignition::math::Vector3d &_scale) override;
+
+      // Documentation inherited.
+      public: virtual MaterialPtr Material() const;
+
+      // Documentation inherited.
+      public: virtual void SetMaterial(
+        MaterialPtr _material, bool _unique) override;
+
+      /// \brief Set material to grid geometry.
```
Comment (ahcorde) [inline ogre2/include/ignition/rendering/ogre2/Ogre2InertiaVisual.hh:68]: review comment

File: ogre2/src/Ogre2InertiaVisual.cc [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+class ignition::rendering::Ogre2InertiaVisualPrivate
+{
+  /// \brief inertia visual materal
+  public: Ogre2MaterialPtr material = nullptr;
+
+  /// \brief Ogre renderable used to render the cross lines.
+  public: std::shared_ptr<Ogre2DynamicRenderable> crossLines = nullptr;
+
+  /// \brief Box visual
+  VisualPtr boxVis = nullptr;
```
Comment (ahcorde) [inline ogre2/src/Ogre2InertiaVisual.cc:41]: public ?

File: ogre2/src/Ogre2InertiaVisual.cc [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+//////////////////////////////////////////////////
+void Ogre2InertiaVisual::Init()
+{
+  BaseInertiaVisual::Init();
+}
+
+//////////////////////////////////////////////////
+void Ogre2InertiaVisual::Load(const ignition::math::Pose3d &_pose,
+                             const ignition::math::Vector3d &_scale)
```
Comment (ahcorde) [inline ogre2/src/Ogre2InertiaVisual.cc:68]:
```suggestion
                              const ignition::math::Vector3d &_scale)
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
Some small comments in the code. Two more things:

  - Can you add some tests to these new classes?
  - Can you add an example in the example folder? I'm fine doing this in another PR and maybe include in the same example wireframes too. And potencially all the other new features.

**Comment by scpeters** [comment]
I was expecting to see a call to [ignition::math::MassMatrix3::EquivalentBox](https://github.com/ignitionrobotics/ign-math/blob/ignition-math6_6.8.0/include/ignition/math/MassMatrix3.hh#L1001-L1012) as we do in osrf/gazebo's [InertiaVisual::Load](https://github.com/osrf/gazebo/blob/gazebo11/gazebo/rendering/InertiaVisual.cc#L53-L75). How do you currently set the size of the inertia box?

**Comment by atharva-18** [comment]
@scpeters The size of the inertia box is currently set in ign-gazebo's [SceneManager](https://github.com/atharva-18/ign-gazebo/blob/visualize_inertia/src/rendering/SceneManager.cc#L1180-L1197).

**Review by ahcorde — APPROVED** [review]
Thank you for adding the tests! Do you mind to change the status of the PR from Draft to a normal PR ?

**Comment by atharva-18** [comment]
I have added a sample in the examples folder
views_demo

File: examples/views_demo/GlutWindow.cc [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+  double x = sin(angle) * 3.0 + 3.0;
+  double y = cos(angle) * 3.0;
+  for (ir::CameraPtr camera : g_cameras)
+  {
+    camera->SetLocalPosition(x, y, 0.0);
+  }
+
+  g_offset += 0.0005;
+}
+//! [update camera]
```
Comment (ahcorde) [inline examples/views_demo/GlutWindow.cc:86]: Do you plan to write a tutorial ? otherwise remove these kind of traces `//!`

File: examples/views_demo/Main.cc [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+  // create center visual
+  VisualPtr center = _scene->CreateVisual();
+  center->AddGeometry(_scene->CreateSphere());
+  center->SetLocalPosition(3, 0, 0);
+  center->SetLocalScale(0.1, 0.1, 0.1);
+  center->SetMaterial(green);
+  root->AddChild(center);
+
+//! [red material]
```
Comment (ahcorde) [inline examples/views_demo/Main.cc:82]: Remove

File: examples/views_demo/Main.cc [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+  sphere->AddGeometry(_scene->CreateSphere());
+  sphere->SetOrigin(0.0, -0.5, 0.0);
+  sphere->SetLocalPosition(3, -1, 0);
+  sphere->SetLocalRotation(0, 0, 0);
+  sphere->SetLocalScale(1, 1, 1);
+  sphere->SetMaterial(red);
+  sphere->SetWireframe(true);
+  root->AddChild(sphere);
+
+//! [white material]
```
Comment (ahcorde) [inline examples/views_demo/Main.cc:104]: Remove

File: examples/views_demo/Main.cc [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    {
+      CameraPtr camera = createCamera(engineName);
+      if (camera)
+      {
+        cameras.push_back(camera);
+      }
+    }
+    catch (...)
+    {
+      // std::cout << ex.what() << std::endl;
```
Comment (ahcorde) [inline examples/views_demo/Main.cc:196]: Remove

**Review by ahcorde — CHANGES_REQUESTED** [review]
Do you mind to add support for windows ? Take a look to this PR https://github.com/ignitionrobotics/ign-rendering/pull/291

**Comment by scpeters** [comment]
>> I was expecting to see a call to [ignition::math::MassMatrix3::EquivalentBox](https://github.com/ignitionrobotics/ign-math/blob/ignition-math6_6.8.0/include/ignition/math/MassMatrix3.hh#L1001-L1012) as we do in osrf/gazebo's [InertiaVisual::Load](https://github.com/osrf/gazebo/blob/gazebo11/gazebo/rendering/InertiaVisual.cc#L53-L75). How do you currently set the size of the inertia box?

> @scpeters The size of the inertia box is currently set in ign-gazebo's [SceneManager](https://github.com/atharva-18/ign-gazebo/blob/visualize_inertia/src/rendering/SceneManager.cc#L1180-L1197).

If this class doesn't actually take inertial parameters as direct inputs, I'm not sure it makes sense to call this class `InertiaVisual`. Currently it's just a box with 3 orthogonal lines. I think it will be easier to generate inertia visuals in other applications if we put the sizing logic in this class rather than require the downstream application to pass the pose and scale. For example, how about the following API?

~~~
virtual void Load(const ignition::math::Inertiald &_inertial);
~~~

**Comment by atharva-18** [comment]
> > > I was expecting to see a call to [ignition::math::MassMatrix3::EquivalentBox](https://github.com/ignitionrobotics/ign-math/blob/ignition-math6_6.8.0/include/ignition/math/MassMatrix3.hh#L1001-L1012) as we do in osrf/gazebo's [InertiaVisual::Load](https://github.com/osrf/gazebo/blob/gazebo11/gazebo/rendering/InertiaVisual.cc#L53-L75). How do you currently set the size of the inertia box?
> 
> > @scpeters The size of the inertia box is currently set in ign-gazebo's [SceneManager](https://github.com/atharva-18/ign-gazebo/blob/visualize_inertia/src/rendering/SceneManager.cc#L1180-L1197).
> 
> If this class doesn't actually take inertial parameters as direct inputs, I'm not sure it makes sense to call this class `InertiaVisual`. Currently it's just a box with 3 orthogonal lines. I think it will be easier to generate inertia visuals in other applications if we put the sizing logic in this class rather than require the downstream application to pass the pose and scale. For example, how about the following API?
> 
> ```
> virtual void Load(const ignition::math::Inertiald &_inertial);
> ```

I see. I think using `math::inertiald` also makes more semantic sense. SImilar to Gazebo classic, we can call `Load(const ignition::math::Inertiald &_inertial)` which in turn calls `Load(const ignition::math::Pose3d &_pose, const ignition::math::Vector3d &_scale)` to create the visual. Thoughts @ahcorde?

**Comment by ahcorde** [comment]
@iche033  do you mind to take a look ?

**Comment by iche033** [comment]
> I see. I think using math::inertiald also makes more semantic sense. Similar to Gazebo classic, we can call Load(const ignition::math::Inertiald &_inertial) which in turn calls Load(const ignition::math::Pose3d &_pose, const ignition::math::Vector3d &_scale) to create the visual.

That sounds good to me. Looking at the implementation, I think one API suggestion is to call the function `SetInertial(...)` instead of `Load(..)`  to indicate to the user that it is safe for this function to be called multiple times as opposed to be a once only thing on load.

File: src/InertiaVisual_TEST.cc [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+  ASSERT_NE(nullptr, inertiaVisual);
+
+  // check initial values
+  EXPECT_EQ(nullptr, inertiaVisual->BoxVisual());
+
+  ignition::math::Pose3d p(0.0, 1.0, 2.5, 1.0, 0.4, 0.4);
+  ignition::math::Vector3d s(0.5, 0.5, 0.5);
+  inertiaVisual->Load(p, s);
+
+  EXPECT_NE(nullptr, inertiaVisual->BoxVisual());
```
Comment (iche033) [inline src/InertiaVisual_TEST.cc:64]: can you expand the test to include other APIs like `Material` and `OgreObject`?

File: examples/views_demo/GlutWindow.cc [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+ *
+ */
+
+#if __APPLE__
+  #include <OpenGL/gl.h>
+  #include <OpenGL/OpenGL.h>
+  #include <GLUT/glut.h>
+#elif _WIN32
+  #define NOMINMAX
+  #include <windows.h>			/* must include this before GL/gl.h */
```
[the remark this answers is not in the document]
Comment (atharva-18) [reply examples/views_demo/GlutWindow.cc]: Changed in [`2673f1b`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/2673f1bee34ec32e94c98ac21e83e01d00948980)

File: examples/views_demo/GlutWindow.cc [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+  GLXContext g_glutContext;
+  Display *g_glutDisplay;
+  GLXDrawable g_glutDrawable;
+#endif
+
+double g_offset = 0.0;
+
+//////////////////////////////////////////////////
+void updateCameras()
+
```
[the remark this answers is not in the document]
Comment (atharva-18) [reply examples/views_demo/GlutWindow.cc]: Changed in [`5f935e1`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/5f935e1d124bc2c91c65f492806c826e2be93e6a)

File: examples/views_demo/Main.cc [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+ * limitations under the License.
+ *
+ */
+
+#if defined(__APPLE__)
+  #include <OpenGL/gl.h>
+  #include <GLUT/glut.h>
+#elif _WIN32
+  #define NOMINMAX
+  #include <windows.h>			      /* must include this before GL/gl.h */
```
[the remark this answers is not in the document]
Comment (atharva-18) [reply examples/views_demo/Main.cc]: Changed in [`5f935e1`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/5f935e1d124bc2c91c65f492806c826e2be93e6a)

File: examples/views_demo/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,40 @@
+cmake_minimum_required(VERSION 3.10.2 FATAL_ERROR)
+project(ignition-rendering-views-demo)
```
[the remark this answers is not in the document]
Comment (atharva-18) [reply examples/views_demo/CMakeLists.txt]: Changed in [`2673f1b`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/2673f1bee34ec32e94c98ac21e83e01d00948980)

File: include/ignition/rendering/base/BaseInertiaVisual.hh [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#ifndef IGNITION_RENDERING_BASE_BASEINERTIAVISUAL_HH_
+#define IGNITION_RENDERING_BASE_BASEINERTIAVISUAL_HH_
+
+#include <vector>
```
[the remark this answers is not in the document]
Comment (atharva-18) [reply include/ignition/rendering/base/BaseInertiaVisual.hh]: Fixed in [`3456a69`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/3456a69186a0e9dac8ec3692461527259b303b95)

File: src/InertiaVisual_TEST.cc (hunk 13 above)
Comment (atharva-18) [reply src/InertiaVisual_TEST.cc:71]: I think the methods to get `Material` and `OgreObject` are local to Ogre 1 and 2 classes only, similar to the [LightVisual test](https://github.com/ignitionrobotics/ign-rendering/blob/ign-rendering5/src/LightVisual_TEST.cc#L54-L67) and the [Ogre1 LightVisual class](https://github.com/ignitionrobotics/ign-rendering/blob/ign-rendering5/ogre/src/OgreLightVisual.cc#L62-L68)

**Comment by scpeters** [comment]
> > I see. I think using math::inertiald also makes more semantic sense. Similar to Gazebo classic, we can call Load(const ignition::math::Inertiald &_inertial) which in turn calls Load(const ignition::math::Pose3d &_pose, const ignition::math::Vector3d &_scale) to create the visual.
> 
> That sounds good to me. Looking at the implementation, I think one API suggestion is to call the function `SetInertial(...)` instead of `Load(..)` to indicate to the user that it is safe for this function to be called multiple times as opposed to be a once only thing on load.

I see that you added the `SetInertial(...)` method in 3456a69, nice work. I did notice that the content of `OgreInertiaVisual::SetInertial` is identical to `Ogre2InertiaVisual::SetInertial`. Since none of that code is ogre-specific, I would move it to `BaseInertiaVisual<T>::SetInertial` and not override the ogre `SetInertial` methods. What do you think @iche033?

**Comment by ahcorde** [comment]
> Since none of that code is ogre-specific, I would move it to `BaseInertiaVisual<T>::SetInertial` and not override the ogre `SetInertial` methods.

+1

**Comment by atharva-18** [comment]
> I see that you added the `SetInertial(...)` method in [3456a69](https://github.com/ignitionrobotics/ign-rendering/commit/3456a69186a0e9dac8ec3692461527259b303b95), nice work. I did notice that the content of `OgreInertiaVisual::SetInertial` is identical to `Ogre2InertiaVisual::SetInertial`. Since none of that code is ogre-specific, I would move it to `BaseInertiaVisual<T>::SetInertial` and not override the ogre `SetInertial` methods. What do you think @iche033?

I moved the content to `BaseInertiaVisual<T>::SetInertial` from the Ogre methods.

**Comment by atharva-18** [comment]
One possible fix to the Ogre2 bug is to destroy all child visuals whenever `BaseInertiaVisual<T>::Destroy` is called
```
@@ -86,6 +89,17 @@ namespace ignition
       T::Init();
     }
 
+    //////////////////////////////////////////////////
+    template <class T>
+    void BaseInertiaVisual<T>::Destroy()
+    {
+      for (unsigned int i = this->ChildCount(); i > 0; --i)
+      {
+        this->ChildByIndex(i - 1)->Destroy();
+      }
+      T::Destroy();
+    }
+
```

**Comment by iche033** [comment]
what's the ogre2 bug? How about being explicit and just removing the BoxVisual instead of all child visuals

**Comment by atharva-18** [comment]
> what's the ogre2 bug? How about being explicit and just removing the BoxVisual instead of all child visuals

In Ogre2, destroying an InertiaVisual doesn't, in turn, destroy its `BoxVisual`. Destroying the box visual explicitly by `this->dataPtr->boxVis->Destroy()` is giving a segfault in the inertia visual test, specifically when calling [`engine->DestroyScene(scene)`](https://github.com/atharva-18/ign-rendering/blob/inertia_visual/src/InertiaVisual_TEST.cc#L74).

**Comment by ahcorde** [comment]
@atharva-18 how I can reproduce the issue ?

**Comment by atharva-18** [comment]
> @atharva-18 how I can reproduce the issue ?

These are the steps:
1. `cd` to the `visualization_demo` directory and add `inertiaVisual->Destroy()` after [Line 133](https://github.com/ignitionrobotics/ign-rendering/blob/63aabbfd065cc68b0e9846763022ce685e41b794/examples/visualization_demo/Main.cc#L133) in the `Main.cc` file.
2. To produce the `boxVisual` "not getting destroyed" bug, comment out [Ogre2InertiaVisual::Destroy()](https://github.com/ignitionrobotics/ign-rendering/blob/63aabbfd065cc68b0e9846763022ce685e41b794/ogre2/src/Ogre2InertiaVisual.cc#L69-L72) method.
3. Now, if you run the demo using Ogre2, the inertia visual's box remains at the origin, detached from its parent.

If we keep the `Ogre2InertiaVisual::Destroy()` method as it is, the box visual gets destroyed, but the unit test gives a segfault.
One way to avoid the segfault is to not explicitly destroy the box visual,
```
@@ -86,6 +89,17 @@ namespace ignition
       T::Init();
     }
 
+    //////////////////////////////////////////////////
+    template <class T>
+    void BaseInertiaVisual<T>::Destroy()
+    {
+      for (unsigned int i = this->ChildCount(); i > 0; --i)
+      {
+        this->ChildByIndex(i - 1)->Destroy();
+      }
+      T::Destroy();
+    }
+
```

Presumably, this bug is also present in `AxisVisual`.

File: ogre2/src/Ogre2InertiaVisual.cc [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+{
+  BaseInertiaVisual::Init();
+}
+
+//////////////////////////////////////////////////
+void Ogre2InertiaVisual::Destroy()
+{
+  if (this->dataPtr->boxVis != nullptr)
+  {
+    this->dataPtr->boxVis->Destroy();
```
Comment (iche033) [inline ogre2/src/Ogre2InertiaVisual.cc:71]: reset `boxVis` so when `Destroy` is called again, the above line won't be executed.

```
this->dataPtr->boxVis.reset();
```

After that, apply this patch, and see if it fixes the crash for you

```
diff --git a/ogre2/src/Ogre2Node.cc b/ogre2/src/Ogre2Node.cc
index 63a43a93..cff1e3e8 100644
--- a/ogre2/src/Ogre2Node.cc
+++ b/ogre2/src/Ogre2Node.cc
@@ -64,6 +64,8 @@ Ogre::SceneNode *Ogre2Node::Node() const
 //////////////////////////////////////////////////
 void Ogre2Node::Destroy()
 {
+  if (ogreNode)
+    return;
```

File: ogre2/src/Ogre2InertiaVisual.cc [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+  }
+  BaseInertiaVisual::Destroy();
```
Comment (iche033) [inline ogre2/src/Ogre2InertiaVisual.cc:73]: I think this is not needed as we are handling most of the destruction of object ourselves.

Comment (iche033) [inline ogre2/src/Ogre2InertiaVisual.cc:73]: Destroy material:

```
  if (this->dataPtr->material && this->Scene())
  {
    this->Scene()->DestroyMaterial(this->dataPtr->material);
    this->dataPtr->material.reset();
  }
```

File: examples/visualization_demo/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+		${GLUT_LIBRARIES}
+  )
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-declarations")
+endif()
+
+if (NOT APPLE)
+  find_package(GLEW REQUIRED)
+  set(TARGET_THIRD_PARTY_DEPENDS
+    ${TARGET_THIRD_PARTY_DEPENDS}
+    GLEW::glew)
```
Comment (iche033) [inline examples/visualization_demo/CMakeLists.txt:27]: I was not able to build the demo with these lines as it complains about glew

I had to replace it with:

```
  if (WIN32)
    set(TARGET_THIRD_PARTY_DEPENDS
      ${TARGET_THIRD_PARTY_DEPENDS}
      GLEW::glew
    )
  else ()
    set(TARGET_THIRD_PARTY_DEPENDS
      ${TARGET_THIRD_PARTY_DEPENDS}
      GLEW
    )
  endif()
```

Comment (atharva-18) [reply examples/visualization_demo/CMakeLists.txt]: FIxed in [`ca3ac18`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/ca3ac18975bb64431ed2b581d648d6815b86cfb0)

File: ogre2/src/Ogre2InertiaVisual.cc (hunk 19 above)
Comment (atharva-18) [reply ogre2/src/Ogre2InertiaVisual.cc:71]: The crash is fixed, but `UNIT_Scene_TEST` seems to be failing

```
      Start 55: UNIT_Scene_TEST
55/93 Test #55: UNIT_Scene_TEST .......................***Failed    1.25 sec

```

File: ogre2/src/Ogre2InertiaVisual.cc (hunk 20 above)
Comment (atharva-18) [reply ogre2/src/Ogre2InertiaVisual.cc:73]: I am getting this message in the console 
```
visualization_demo: /var/lib/jenkins/workspace/ogre-2.1-debbuilder/repo/OgreMain/src/OgreHlmsDatablock.cpp:170: virtual Ogre::HlmsDatablock::~HlmsDatablock(): Assertion `mLinkedRenderables.empty() && "This Datablock is still being used by some Renderables." " Change their Datablocks before destroying this."' failed.
Aborted (core dumped)
```

I think this is because of `crossLines` still using the material. Explicitly destroying `crossLines` before the material is giving an exception in the test
```
      Start 17: UNIT_InertiaVisual_TEST
17/93 Test #17: UNIT_InertiaVisual_TEST ...............Child aborted***Exception:   0.33 sec
```

Comment (atharva-18) [reply ogre2/src/Ogre2InertiaVisual.cc:73]: Destroying the `crossLines` still requires us to call `BaseInertiaVisual::Destroy()`, see https://github.com/ignitionrobotics/ign-rendering/pull/326#discussion_r649923195

File: ogre2/src/Ogre2InertiaVisual.cc (hunk 19 above)
Comment (iche033) [reply ogre2/src/Ogre2InertiaVisual.cc:71]: oops that should've been `if (!ogreNode)`

Can you take a look at the changes in this commit:

https://github.com/ignitionrobotics/ign-rendering/commit/9575d3ddc0f4524348a4b203e6b1e3a51e33c6a3

I think that fixes the test and crash.

File: examples/visualization_demo/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+    set(TARGET_THIRD_PARTY_DEPENDS
+      ${TARGET_THIRD_PARTY_DEPENDS}
+      GLEW
+    )
+  endif()
+endif()
+
+if (WIN32)
+	find_package(FreeGLUT REQUIRED)
+	set(TARGET_THIRD_PARTY_DEPENDS ${TARGET_THIRD_PARTY_DEPENDS} FreeGLUT::freeglut)
```
Comment (ahcorde) [inline examples/visualization_demo/CMakeLists.txt:40]:
```suggestion
  find_package(FreeGLUT REQUIRED)
  set(TARGET_THIRD_PARTY_DEPENDS ${TARGET_THIRD_PARTY_DEPENDS} FreeGLUT::freeglut)
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: ogre2/src/Ogre2InertiaVisual.cc (hunk 19 above)
Comment (atharva-18) [reply ogre2/src/Ogre2InertiaVisual.cc:71]: Fixed in [`dd42817`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/dd42817dd2b396b738ff0988850e433f41f1cd4a)

File: ogre2/src/Ogre2InertiaVisual.cc (hunk 20 above)
Comment (atharva-18) [reply ogre2/src/Ogre2InertiaVisual.cc]: Fixed in [`dd42817`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/dd42817dd2b396b738ff0988850e433f41f1cd4a)

Comment (atharva-18) [reply ogre2/src/Ogre2InertiaVisual.cc]: Fixed in [`dd42817`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/dd42817dd2b396b738ff0988850e433f41f1cd4a)

File: examples/custom_shaders_uniforms/CMakeLists.txt [hunk 23]
```diff
@@ -5,7 +5,7 @@ include_directories(SYSTEM
   ${PROJECT_BINARY_DIR}
 )
 
-find_package(ignition-rendering4)
+find_package(ignition-rendering6)
```
Comment (ahcorde) [inline examples/custom_shaders_uniforms/CMakeLists.txt:8]: This change is not related to this PR I prefer to remove it from this PR

**Review by ahcorde — CHANGES_REQUESTED** [review]
Last comment from me

File: examples/custom_shaders_uniforms/CMakeLists.txt (hunk 23 above)
Comment (atharva-18) [reply examples/custom_shaders_uniforms/CMakeLists.txt]: Removed the commit - https://github.com/ignitionrobotics/ign-rendering/compare/617af60fdaa475ec13612f41f1068d574df12b34..242888265c0c2f5cd4ac9a5612e0015fcfe65609

**Review by ahcorde — APPROVED** [review]

File: examples/custom_shaders_uniforms/CMakeLists.txt (hunk 23 above)
Comment (iche033) [reply examples/custom_shaders_uniforms/CMakeLists.txt]: created PR for this change: https://github.com/ignitionrobotics/ign-rendering/pull/343

File: src/base/BaseScene.cc [hunk 24]
```diff
@@ -1350,6 +1381,15 @@ void BaseScene::CreateMaterials()
...
   material->SetLightingEnabled(false);
 
+  material = this->CreateMaterial("Default/TransPurple");
+  material->SetAmbient(1.0, 0.0, 1.0);
+  material->SetDiffuse(1.0, 0.0, 1.0);
+  material->SetEmissive(1.0, 0.0, 1.0);
+  material->SetTransparency(0.5);
+  material->SetCastShadows(false);
+  material->SetReceiveShadows(false);
+  material->SetLightingEnabled(false);
```
Comment (iche033) [inline src/base/BaseScene.cc:1391]: looks like the semi-transparent purple box is occluding the green lines in the ogre2 visualization. I think you can disable depth write to make the transparency effect more like ogre 1.x (which has depth write disabled by default when objects are semi-transparent).

```
material->SetDepthWriteEnabled(false);
```

**Review by iche033 — COMMENTED** [review]
just have one more comment about ogre2 transparency setting for the purple box, otherwise it's working well for me.

File: src/base/BaseScene.cc (hunk 24 above)
Comment (atharva-18) [reply src/base/BaseScene.cc:1391]: Fixed in [`42a6a414`](https://github.com/ignitionrobotics/ign-rendering/pull/326/commits/42a6a4143d1a6729a3d1b8a2b54e3d23d8970669)

**Review by iche033 — APPROVED** [review]

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/google-summer-of-code-2021-new-gui-widgets-in-ignition-gazebo/1081/1
